src/util/result.h

```diff
@@ -31,54 +148,84 @@ struct Error {
-    Result(T obj) : m_variant{std::in_place_index_t<1>{}, std::move(obj)} {}
-    Result(Error error) : m_variant{std::in_place_index_t<0>{}, std::move(error.message)} {}
+    template <typename Fn, typename... Args>
+    void Construct(const Fn& fn, Warning warning, Args&&... args)
+    {
+        this->AddWarning(std::move(warning.message));
+        Construct(fn, std::forward<Args>(args)...);
+    }
 
-    //! std::optional methods, so functions returning optional<T> can change to
-    //! return Result<T> with minimal changes to existing code, and vice versa.
-    bool has_value() const noexcept { return m_variant.index() == 1; }
-    const T& value() const LIFETIMEBOUND
+    template <typename Fn, typename OT, typename OF, typename... Args>
+    void Construct(const Fn& fn, Result<OT, OF>&& other, Args&&... args)
+    {
+        *this << std::move(other);
+        Construct(fn, std::forward<Args>(args)...);
+    }
+
+    template <typename OT, typename OF>
+    void MoveConstruct(Result<OT, OF>& other)
     {
-        assert(has_value());
-        return std::get<1>(m_variant);
+        *this << std::move(other);
+        if (other) this->MoveValue(other); else if (other.m_info) this->m_info->failure = std::move(other.m_info->failure);
     }
-    T& value() LIFETIMEBOUND
+
+    template <typename FT, typename FF>
+    friend class Result;
+
+public:
+    //! Constructors, destructor, and assignment operator.
+    template <typename... Args>
+    Result(Args&&... args)
     {
-        assert(has_value());
-        return std::get<1>(m_variant);
+        Construct([this](auto&&... x) { this->InitValue(std::forward<decltype(x)>(x)...); }, std::forward<Args>(args)...);
     }
-    template <class U>
-    T value_or(U&& default_value) const&
+    template <typename OT, typename OF>
+    Result(Result<OT, OF>&& other) { MoveConstruct(other); }
+    Result& operator=(Result&& other)
     {
-        return has_value() ? value() : std::forward<U>(default_value);
+        if (*this) this->DestroyValue(); else this->m_info->failure.reset();
+        MoveConstruct(other);
+        return *this;
     }
-    template <class U>
-    T value_or(U&& default_value) &&
+    ~Result() { if (*this) this->DestroyValue(); }
+
+    //! Operator moving warning and error messages from one result to another.
+    template <typename OT, typename OF>
+    Result<OT, OF>&& operator<<(Result<OT, OF>&& other)
     {
-        return has_value() ? std::move(value()) : std::forward<U>(default_value);
+        if (other.m_info && !this->m_info) {
+            this->m_info.reset(new detail::ErrorInfo<F>{.errors = std::move(other.m_info->errors),
+                                                        .warnings = std::move(other.m_info->warnings)});
+        } else if (other.m_info && this->m_info) {
+            detail::MoveElements(other.m_info->errors, this->m_info->errors);
+            detail::MoveElements(other.m_info->warnings, this->m_info->warnings);
+        }
```

Comment:
What is the difference between when the `this->m_info` is `true` and when it is `false`? While as far as I understand in both cases the `this->m_info` is destroyed and replaced by `other.m_info`.  Am I correct? So why bother to check `other.m_info && !this->m_info` and  `other.m_info && this->m_info`?

Reply:
> What is the difference between when the `this->m_info` is `true` and when it is `false`?

The implementation is optimized for the "happy path" where a success value is set and there is no failure value and no error or warning message strings. In the happy path case, `m_info` is null and no allocation is needed. Otherwise, if there has been any kind of error or warning `m_info` will be non-null.

> While as far as I understand in both cases the `this->m_info` is destroyed and replaced by `other.m_info`.  Am I correct?

No that's not really correct. As the comment "Operator moving warning and error messages from one result to another" says, only warning and error strings are moved from `other` to `*this`. The success and value statuses of `other` and `*this` objects are unchanged, and the success and failure values of both objects (if any) are also unchanged. Existing warning and error strings in `*this` are also not affected, the new errors and warnings just get appended and don't replace existing one.

> So why bother to check `other.m_info && !this->m_info` and `other.m_info && this->m_info`?

I'm not sure what other code you might expect to see here, but in both cases this is just moving `other.m_info->errors` strings to `this->m_info->errors` and moving `other.m_info->warnings` strings to `this->m_info->warnings` while not deleting any existing strings and not changing `other.m_info->failure` and to `this->m_info->failure` values.

Reply:
Thanks for your explanation. My wrong assumption was that by replacing the warnings and errors the `this` is completely like the `other`, so why checking whether `this->m_info` is null or not. 
But as you mentioned the success and value statuses remained unchanged.